### PR TITLE
Login via QR code

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/0/1204039315134727/f 

### Description
This allow users to perform login reading qr code or generate qr to perform login from another device.

### Steps to test this PR

_Feature 1_ (you will need 2 devices here)

From device A
- [x] go to sync internal settings
- [x] Create account
- [x] click show qr code
From device B (should be unauthenticated)
- [x] go to sync internal settings
- [x] click read qr code and read qr code in device A
- [x] ensure login works

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #2906** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/2906" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/2906?utm_source=stack-comment).

File: sync/sync-impl/build.gradle [hunk 1]
```diff
@@ -44,6 +44,8 @@ dependencies {
     implementation "com.squareup.moshi:moshi-kotlin:_"
     implementation Google.dagger
     implementation AndroidX.core.ktx
+    implementation('com.journeyapps:zxing-android-embedded:4.3.0') { transitive = false }
```
Comment (malmstein) [inline sync/sync-impl/build.gradle:47]: should the version be in the `versions.properties` file?

File: sync/sync-impl/build.gradle [hunk 2]
```diff
@@ -44,6 +44,8 @@ dependencies {
...
+    implementation 'com.google.zxing:core:3.3.0'
```
Comment (malmstein) [inline sync/sync-impl/build.gradle:48]: ditto

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 3]
```diff
@@ -25,18 +25,25 @@ import com.duckduckgo.sync.store.SyncStore
...
 import com.squareup.moshi.Moshi
-import javax.inject.Inject
+import javax.inject.*
+import kotlin.DeprecationLevel.WARNING
 import timber.log.Timber
 
 interface SyncRepository {
     fun createAccount(): Result<Boolean>
+
+    @Deprecated(message = "Method only used for testing purposes. Relies on a local stored recovery key.", level = DeprecationLevel.WARNING)
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:35]: can you make this internal then?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 4]
```diff
@@ -25,18 +25,25 @@ import com.duckduckgo.sync.store.SyncStore
...
     fun login(): Result<Boolean>
+    fun login(recoveryCodeRawJson: String): Result<Boolean>
     fun getAccountInfo(): AccountInfo
+
+    @Deprecated(message = "Method only used for testing purposes.", level = DeprecationLevel.WARNING)
     fun storeRecoveryCode()
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:41]: ditto

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 5]
```diff
@@ -211,9 +237,13 @@ data class AccountInfo(
...
 )
 
+data class LinkCode(
+    val recovery: RecoveryCode,
+)
+
 data class RecoveryCode(
-    val primaryKey: String,
-    val userID: String,
+    val primary_key: String,
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:245]: I really don't like to leak the API terrible field names into our code.

```suggestion
  @Json(name = "primary_key) val primaryKey: String
```

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 6]
```diff
@@ -211,9 +237,13 @@ data class AccountInfo(
...
+    val user_id: String,
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:246]: ditto

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt [hunk 7]
```diff
@@ -36,6 +47,17 @@ class SyncInitialSetupActivity : DuckDuckGoActivity() {
     private val binding: ActivitySyncSetupBinding by viewBinding()
     private val viewModel: SyncInitialSetupViewModel by bindViewModel()
 
+    // Register the launcher and result handler
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt:50]: nit: unnecessary

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt [hunk 8]
```diff
@@ -71,12 +97,37 @@ class SyncInitialSetupActivity : DuckDuckGoActivity() {
...
             }
+
+            ReadQR -> {
+                barcodeLauncher.launch(getScanOptions())
+            }
+
+            is ShowQR -> {
+                try {
+                    val barcodeEncoder = BarcodeEncoder()
+                    val bitmap: Bitmap = barcodeEncoder.encodeBitmap(command.string, QR_CODE, 400, 400)
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt:111]: shouldn't this be done in a background thread?

**Review by malmstein — APPROVED** [review]
Couple nits, nothing blocking.

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt (hunk 3 above)
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:35]: I assume you mean `internal` visibility.

The repository itself is not exposed, so it's only used inside the :impl module. Adding `internal` will make no difference in this case.

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt (hunk 8 above)
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupActivity.kt:111]: not sure 100%, imho we don't need to move into background here, but happy to change it if you see a reason. I will move forward with how it is now, but I can do it in the final screens.

**Pull request merged**